src/CapabilityService.WebApi/Infrastructure/Persistence/CapabilityTransactionalDecorator.cs

```diff
@@ -172,6 +31,16 @@ public async Task<Capability> UpdateCapability(Guid capabilityId, string newName
 
         public Task<IEnumerable<Capability>> GetAllCapabilities() => _inner.GetAllCapabilities();
         public Task<Capability> GetCapability(Guid id) => _inner.GetCapability(id);
+        public async Task DeleteCapability(Guid id)
+        {
+            using (var transaction = await _dbContext.Database.BeginTransactionAsync())
```

Comment:
We only call SaveChanges once in the scope of this transaction...

Reply:
I am not sure we can expect a TransactionalDecorator to know if the wrapped classes has preformed the SaveChanges or not

Reply:
Due to the way we have chosen to implement the "decorator pattern" (using only interfaces and no shared abstractions) we end up being forced to implement all the interface methods and in doing so it looks like we felt a need to add logic to make our "golden hammer" look good, when we should probably be using a different hammer called a "unit-of-work" pattern (https://docs.microsoft.com/en-us/aspnet/mvc/overview/older-versions/getting-started-with-ef-5-using-mvc-4/implementing-the-repository-and-unit-of-work-patterns-in-an-asp-net-mvc-application) to solve this particular challenge, as it can control transactions that span multiple entities/repos. We can also just do what most people I know do these days and accept that the DbContext itself is already kind of a unit of work with a Complete method called SaveChanges and if we never call SaveChanges in our repos then we end up saving a shit ton of code while obtaining the same thing MS does with their sample code. :)

Reply:
I agree the decorator pattern gets cumbersome when you have no shared abstractions.

I feel the original author had a good idea with the explicitness of the transaction. I also feel that DbContext does a great job of keeping track of changes.

I have a inking that the idea was allow SaveChanges to be called multiple places and tie them all back to the same transaction. Instead of what we currently have where SaveChanges basically only happens in the same method as the transaction.

 `Task have been added to the todo list`